source/GameAction.cpp

```diff
@@ -0,0 +1,223 @@
+				out.Write(line);
+		}
+		out.EndChild();
+	}
+	for(const auto &it : specialLogText)
+		for(const auto &eit : it.second)
+		{
+			out.Write("log", it.first, eit.first);
+			out.BeginChild();
+			{
+				// Break the text up into paragraphs.
+				for(const string &line : Format::Split(eit.second, "\n\t"))
+					out.Write(line);
+			}
+			out.EndChild();
+		}
+	if(payment)
+		out.Write("payment", payment);
+	for(const auto &it : events)
+	{
+		if(it.second.first == it.second.second)
+			out.Write("event", it.first->Name(), it.second.first);
+		else
+			out.Write("event", it.first->Name(), it.second.first, it.second.second);
+	}
+	for(const auto &name : fail)
+		out.Write("fail", name);
+	
+	conditions.Save(out);
+}
+
+
+
+bool GameAction::IsEmpty() const
+{
+	return empty;
+}
+
+
+
+int GameAction::Payment() const
+{
+	return payment;
+}
+
+
+
+// Do the actions of the GameAction.
+void GameAction::Do(PlayerInfo &player) const
+{
+	if(!logText.empty())
+		player.AddLogEntry(logText);
+	for(const auto &it : specialLogText)
+		for(const auto &eit : it.second)
+			player.AddSpecialLog(it.first, eit.first, eit.second);
+	
+	if(payment)
+	{
+		// Conversation actions don't block a mission from offering if a
+		// negative payment would drop the player's account balance below
+		// zero, so negative payments need to be handled.
+		int64_t account = player.Accounts().Credits();
+		// If the payment is negative and the player doesn't have enough
+		// in their account, then the player's credits are reduced to 0.
+		if(account + payment >= 0)
+			player.Accounts().AddCredits(payment);
+		else if(account > 0)
+			player.Accounts().AddCredits(-account);
+	}
+	
+	for(const auto &it : events)
+		player.AddEvent(*it.first, player.GetDate() + it.second.first);
+	
+	if(!fail.empty())
+	{
+		// If this action causes this or any other mission to fail, mark that
+		// mission as failed. It will not be removed from the player's mission
+		// list until it is safe to do so.
+		for(const Mission &mission : player.Missions())
+			if(fail.count(mission.Identifier()))
+				player.FailMission(mission);
+	}
+	
+	// Check if applying the conditions changes the player's reputations.
+	player.SetReputationConditions();
+	conditions.Apply(player.Conditions());
+	player.CheckReputationConditions();
+}
+
+
+
+GameAction GameAction::Instantiate(map<string, string> &subs, int jumps, int payload) const
+{
+	GameAction result;
+	
+	result.empty = empty;
+	
+	for(const auto &it : events)
+	{
+		// Allow randomization of event times. The second value in the pair is
+		// always greater than or equal to the first, so Random::Int() will
+		// never be called with a value less than 1.
+		int day = it.second.first + Random::Int(it.second.second - it.second.first + 1);
+		result.events[it.first] = make_pair(day, day);
+	}
+	result.payment = payment + (jumps + 1) * payload * paymentMultiplier;
+	// Fill in the payment amount if this is the "complete" action.
+	if(result.payment)
```

Comment:
How do we ensure that filling out the payment is only done for the complete action? We should not overwrite payment with an in-conversation payment.

Reply:
Btw, if we take this out of the Instantiate and just handle this outside this function for the "on complete" section only, then we can make the subs map const again.

Reply:
This is another instance of this simply being something that was a part of the MissionAction, but I think we might want to keep this here. I mentioned how you can't use a \<payment> text replacement to get the payment from an action, but if subs is non-const, then an action node can change the payment replacement for the text nodes after it.

Reply:
The thing is, if I'm in the middle of a conversation, then I might want to refer both to the payment on the end (on complete) and to the payment in a conversation node; for example in a sentence like:
"Listen, I have a job for you, you get 5000 credits upfront to buy some weapons, and you will get <payment> if you deliver the weapons to the shady figure waiting for you at <destination>."

Reply:
As I understand it, the current behavior is as follows:
* `subs` is created in Mission::Instantiate and filled in with various substitutions that don't get changed.
* NPCs and their conversations/dialogs get instantiated.
* The `on complete` conversation gets instantiated. This fills in the `<payment>` and `<fine>` subs.
  * NPC conversations can't use the `<payment>` and `<fine>` subs because of this? (This could be fixed by cherry picking the payment and fine out of the `on complete` action before doing NPC or MissionAction instantiations, as just reversing the order of instantiations would mean that MissionActions couldn't use the `<npc>` sub.)
* The rest of the `on *` actions are instantiated. Each action, if it has its own payment or fine, replaces the `<payment>` or `<fine>` substitutions with its own values before instantiating its conversation/dialog, then replaces the subs with the old values. This means that any `on *` action with a payment can't refer to the `on complete` mission payment, but other `on *` actions that don't have a payment could.
* NEW BEHAVIOR: if a conversation action has a payment or fine, it will replace the `<payment>` or `<fine>` subs. This means that you can not refer to any `on *` action payment after a conversation action has replaced the payment. The `on complete` subs for `<payment>` and `<fine>` are still restored after the conversation ends.

So we have three types of payments here that we might want to refer to: the `on complete` payment, the current `on *` payment, and the most recent `action` payment.

Reply:
I've opened issue #6258 for tracking this discussion. This issue would also apply to the new payment method added by conversation actions.

Reply:
Fine for me to keep this for a follow-up PR and track this in #6258.